### PR TITLE
Fix #86 - catch "Trying to save highlighting twice"

### LINKED ISSUE #86: sonar-yaml 1.8.0 fails on k8s YAML files (Trying to save highlighting twice)
With `sonar-yaml` 1.8.0, in a project which contains k8s YAML files, analysis fails with the following exception:
```
INFO: ------------------------------------------------------------------------
INFO: EXECUTION FAILURE
INFO: ------------------------------------------------------------------------
INFO: Total time: 15.860s
INFO: Final Memory: 39M/148M
INFO: ------------------------------------------------------------------------
ERROR: Error during SonarScanner execution
java.lang.UnsupportedOperationException: Trying to save highlighting twice for the same file is not supported: k8s.yml
	at org.sonar.scanner.sensor.DefaultSensorStorage.store(DefaultSensorStorage.java:303)
	at org.sonar.api.batch.sensor.highlighting.internal.DefaultHighlighting.doSave(DefaultHighlighting.java:109)
	at org.sonar.api.batch.sensor.internal.DefaultStorable.save(DefaultStorable.java:45)
	at com.github.sbaudoin.sonar.plugins.yaml.rules.YamlSensor.saveSyntaxHighlighting(YamlSensor.java:314)
	at com.github.sbaudoin.sonar.plugins.yaml.rules.YamlSensor.execute(YamlSensor.java:120)
	at org.sonar.scanner.sensor.AbstractSensorWrapper.analyse(AbstractSensorWrapper.java:64)
```
(that's easy to reproduce with a single file, mine is the example from [S6428](https://sqaas.dos.tech.orange/coding_rules?languages=kubernetes&open=kubernetes%3AS6428))

Before 1.8.0, `sonar-yaml` was failing already, except a bit earlier, on this other conflict which is now catched and turned into a warning (#70):
```
INFO: Sensor YAML Sensor [yaml]
WARN: Cannot save measures for file k8s.yml, ignoring them
java.lang.UnsupportedOperationException: Can not add the same measure twice on k8s.yml: DefaultMeasure[component=k8s.yml,metric=Metric[uuid=<null>,key=ncloc_data,description=<null>,type=DATA,direction=0,domain=<null>,name=ncloc_data,qualitative=false,userManaged=false,enabled=true,worstValue=<null>,bestValue=<null>,optimizedBestValue=false,hidden=false,deleteHistoricalData=false,decimalScale=<null>],value=1=1;2=1;3=1;4=1;5=1;6=1;7=1;8=1;9=1;10=1;11=1;12=1;13=1;14=1,fromCore=false,storage=org.sonar.scanner.sensor.DefaultSensorStorage@4785f176,saved=false]
	at org.sonar.scanner.sensor.DefaultSensorStorage.saveMeasure(DefaultSensorStorage.java:179)
	at org.sonar.scanner.sensor.DefaultSensorStorage.store(DefaultSensorStorage.java:139)
	at org.sonar.api.batch.sensor.measure.internal.DefaultMeasure.doSave(DefaultMeasure.java:96)
	at org.sonar.api.batch.sensor.internal.DefaultStorable.save(DefaultStorable.java:45)
	at org.sonar.scanner.DefaultFileLinesContext.save(DefaultFileLinesContext.java:97)
	at com.github.sbaudoin.sonar.plugins.yaml.linecounter.LineCounter.saveMeasures(LineCounter.java:78)
	at com.github.sbaudoin.sonar.plugins.yaml.linecounter.LineCounter.analyse(LineCounter.java:56)
	at com.github.sbaudoin.sonar.plugins.yaml.rules.YamlSensor.computeLinesMeasures(YamlSensor.java:158)
	at com.github.sbaudoin.sonar.plugins.yaml.rules.YamlSensor.execute(YamlSensor.java:119)
	at org.sonar.scanner.sensor.AbstractSensorWrapper.analyse(AbstractSensorWrapper.java:64)
```

I will propose a merge-request to:
- similarly catch the `UnsupportedOperationException: Trying to save highlighting twice` and continue despite the error, not saving highlighting
- add a k8s YAML file to integration tests (generates 2 warnings in the scanner log, which are expected and thus ignored)
- be less verbose on the `WARN: Cannot save measures for...` (only print full exception if in debug mode, otherwise it becomes noisy/scary when analysing a project with many such errors)

### PR DESCRIPTION
**Pull request by thomasgl-orange**
Fix #86.

- catch the `UnsupportedOperationException: Trying to save highlighting twice` on k8s files, and continue despite the error, not saving highlighting
- add a k8s YAML file to integration tests (generates 2 warnings in the scanner log, which are expected and thus ignored); the example comes from [S6428](https://sqaas.dos.tech.orange/coding_rules?languages=kubernetes&open=kubernetes%3AS6428)
- be less verbose on the `WARN: Cannot save measures for...` (only print full exception if in debug mode, otherwise it becomes noisy/scary when analysing a project with many such errors)

### CODE REVIEW

**Comment by thomasgl-orange** [comment]
Got an error on Travis which I couldn't reproduce in my dev environment: https://app.travis-ci.com/github/sbaudoin/sonar-yaml/builds/267178582#L424
```
Tests run: 11, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 3.711 sec <<< FAILURE! - in com.github.sbaudoin.sonar.plugins.yaml.rules.YamlSensorTest
testGlobalConfig0(com.github.sbaudoin.sonar.plugins.yaml.rules.YamlSensorTest)  Time elapsed: 0.02 sec  <<< FAILURE!
java.lang.AssertionError: expected null, but was:<com.github.sbaudoin.yamllint.YamlLintConfig@1af1aae9>
	at com.github.sbaudoin.sonar.plugins.yaml.rules.YamlSensorTest.testGlobalConfig0(YamlSensorTest.java:153)
```
I don't think it's related to this MR. Anyway, not sure if it was the proper fix or pure luck, but I've added 423a0b1 and now the unit tests are okay: https://app.travis-ci.com/github/sbaudoin/sonar-yaml/builds/267179010 (integration tests still running)

File: it/audit.sh [hunk 1]
```diff
@@ -22,7 +22,10 @@ fi
 
 # Check for warnings
 # Ugly fix for SQ 9.8+ that has deprecated login/password authentication but still allows it.
-if grep -iv "propert.* 'sonar.password' .* deprecated" /tmp/scanner.log | grep -q "^WARN: "
+# Also ignore expected warnings regarding k8s.yaml
+if grep -iv "propert.* 'sonar.password' .* deprecated" /tmp/scanner.log \
+ | grep -v "Cannot save .* k8s.yml, ignoring" \
```
Comment (sbaudoin) [inline it/audit.sh:27]: Implemented this way, I'm not sure this is relevant. Better put the `k8s.yml` file as a unit test to check the `UnsupportedOperationException`.

Comment (sbaudoin) [reply it/audit.sh:27]: Thanks for the update.

The purpose of unit tests is to make sure things are working as expected. This is valid until you discover something... unexpected, in which case you have to provide a fix and incorporate the unexpected as now something you expect.

Anyway, we may keep `k8s.yml` in `it/src`, there is nothing wrong with that.

**Comment by thomasgl-orange** [comment]
Added [9e1f63c](https://github.com/sbaudoin/sonar-yaml/pull/87/commits/9e1f63cbdb1efe3c54c396322f222b08de2e6cb1) to fix random failure of one unit test: depending on tests order, in `YamlSensorTest.testGlobalConfig0()`, the `user.home` system property might have been previously set to a directory which contains a yamllint configuration, and left as-is.

**Comment by thomasgl-orange** [comment]
FYI, we've successfully tested a custom build of 1.8.0 + this PR on our SQ 10.2.1 **Enterprise** server. No additional incompatibilities detected, analysis of projects with YAML files, including k8s-related ones, works fine.

**Comment by LomeshDaheria** [comment]
Hi Thomas,
When can we expect to get the new version of sonar-yaml plugin 1.8.0+ as version below 1.8.0 is causing this error
**Can not add the same measure twice on xyz.yaml**
while version 1.8.0 is causing
**Trying to save highlighting twice for the same file is not supported in xyz.yaml**
hoping it to use very soon.
Thanks in advance

**Comment by sbaudoin** [comment]
I will certainly do rebase and comment rewrite before merging this PR. I will accept https://github.com/sbaudoin/sonar-yaml/pull/87/commits/9e1f63cbdb1efe3c54c396322f222b08de2e6cb1 but temporarily: Junit 5 provides try-with-resources that will in all likelihood fix this in a "elegant" way.

**Comment by sbaudoin** [comment]
@LomeshDaheria will try to release a fixed version today. The JAR will be available in GitHub before it is available in the SonarQube marketplace.

**Comment by sbaudoin** [comment]
> Added [9e1f63c](https://github.com/sbaudoin/sonar-yaml/pull/87/commits/9e1f63cbdb1efe3c54c396322f222b08de2e6cb1) to fix random failure of one unit test: depending on tests order, in `YamlSensorTest.testGlobalConfig0()`, the `user.home` system property might have been previously set to a directory which contains a yamllint configuration, and left as-is.

Had to roll it back: it causes the build to fail.

**Comment by sbaudoin** [comment]
Version 1.8.1 released and [available on GitHub](https://github.com/sbaudoin/sonar-yaml/releases/tag/v1.8.1)

**Comment by thomasgl-orange** [comment]
Thanks for the release @sbaudoin.

> Had to roll it back: it causes the build to fail.

Oops... for my defense, I swear Travis build was green after I've pushed [9e1f63c](https://github.com/sbaudoin/sonar-yaml/pull/87/commits/9e1f63cbdb1efe3c54c396322f222b08de2e6cb1) (and it was indeed fixing a reproducible test failure, for the reason I've explained).

I think the explanation for the later build breaking is [here](https://stackoverflow.com/questions/68791709/paths-get-path-of-throws-nullpointerexception-on-upgrade-to-java-16). On JDK 11, `Paths.get()` (used in `getUserGlobalConfigPath()`) was happily swallowing a `null` root path, and this has been fixed in later version. So it started leading to an NPE once merged on top of 58e6a6bc6156211afc05e0511f8d5db699f4f6db (JDK 17 upgrade).

I guess I should have restored some default value for `user.home`, instead of clearing it. It crossed my mind TBH, but I went with the shortest tests-passing version.

**Comment by sbaudoin** [comment]
Never mind, things are working now. I'm going to try to upgrade to JUnit 5 so I'll optimize some of the tests on the way.

**Pull request merged**